src/common/offline_transformations/src/pruning/propagate_masks.cpp

```diff
@@ -83,37 +100,76 @@ class ngraph::pass::mask_propagation::MatMul : public MatcherPass {
-                if (a_mask_row->at(a_inner_dim) != b_mask_row->at(b_inner_dim))
-                    cur_mask->initialize_dependencies();
-                return true;
-            };
             // Connect a with matmul mask
-            matmul_mask->add_callback(matmul_callback, a_mask);
-            a_mask->add_callback([=](Mask::Ptr cur_mask) -> bool {
-                cur_mask->at(a_inner_dim) = b_mask_row->at(b_inner_dim);
-                cur_mask->at(a_outer_dim) = matmul_mask_row->at(matmul_rows_dim);
-                return true;
-            }, matmul_mask);
+            if (a_mask) {
+                bool init = true;
+                a_mask->add_callback([=](Mask::Ptr cur_mask) -> bool {
+                    auto result_mask = std::make_shared<Mask>(cur_mask->size());
+                    result_mask->copy_value_from_mask_reversed(matmul_mask_row);
+                    result_mask->at(a_inner_dim) = b_mask_row->at(b_inner_dim);
+                    result_mask->at(a_outer_dim) = matmul_mask_row->at(matmul_rows_dim);
+                    cur_mask->copy_value_from_mask(result_mask.get());
+                    return true;
+                }, matmul_mask);
+                matmul_mask->add_callback([=](Mask::Ptr cur_mask) mutable -> bool {
+                    auto result_mask = std::make_shared<Mask>(cur_mask->size());
+                    result_mask->copy_value_from_mask(cur_mask.get());
+                    result_mask->copy_value_from_mask_reversed(a_mask_row);
+                    if (init) {
+                        result_mask->at(matmul_cols_dim) = b_mask_row->at(b_outer_dim);
+                        init = false;
+                    } else {
+                        result_mask->at(matmul_cols_dim) = cur_mask->at(matmul_cols_dim);
+                    }
+                    result_mask->at(matmul_rows_dim) = a_mask_row->at(a_outer_dim);
+                    if (a_mask_row->at(a_inner_dim) != b_mask_row->at(b_inner_dim))
+                        cur_mask->initialize_dependencies();
+                    cur_mask->copy_value_from_mask_reversed(result_mask.get());
+                    return true;
+                }, a_mask);
+            }
             // connect b with matmul mask
-            matmul_mask->add_callback(matmul_callback, b_mask);
             b_mask->add_callback([=](Mask::Ptr cur_mask) -> bool {
-                cur_mask->at(b_inner_dim) = a_mask_row->at(a_inner_dim);
-                cur_mask->at(b_outer_dim) = matmul_mask_row->at(matmul_cols_dim);
+                auto result_mask = std::make_shared<Mask>(cur_mask->size());
+                result_mask->copy_value_from_mask_reversed(matmul_mask_row);
+                if (a_mask_row)
+                    result_mask->at(b_inner_dim) = a_mask_row->at(a_inner_dim);
+                else
+                    result_mask->at(b_inner_dim).clear();
+                result_mask->at(b_outer_dim) = matmul_mask_row->at(matmul_cols_dim); // TODO: remove this line
```

Comment:
do we still need remove this line?

Reply:
Oh, this comment like
"It's possible that this line isn't affect result mask in any way, but I'm not 100% sure"
I'm out of context for now unfortunately, so I suggest to keep this comment and this line till the moment we will prof this line is doing nothing to result mask.

Reply:
Besides we have comment "think about it" in `shrink_wegihts.cpp` leaved by Gleb, which means
"This condition is always false, but I'm not 100% sure, so let's keep this" :)